File: javascript/binary/binary.js
'use strict';

class Binary {
  constructor(digits) {
    this.digits = stringToDigitList(digits);
  }

  toDecimal() {
    return binaryListToDecimal(this.digits, 0);
  }
}

function stringToDigitList(digitString) {
  if (digitString.match(/[^0|1]/)) { return []; }

  return digitString.split('').map((d) => parseInt(d, 10));
}

function binaryListToDecimal(digits) {
  return digits.reduce(addBinaryDigits, 0);
}

function addBinaryDigits(sum, digit, index, digits) {
  if (digit === 1) {
    let power = digits.length - index - 1;
    sum = sum + Math.pow(2, power);
  }
  return sum;
}

module.exports = Binary;

File: ruby/binary/binary.rb
class Binary
  def self.to_decimal(input)
    raise ArgumentError, "Input must contain only 0 and 1" unless input.match(/^[01]+$/)

    input
      .split('')
      .reverse
      .each_with_index
      .map { |digit, factor| digit.to_i * 2**factor }
      .reduce(&:+)
  end
end

module BookKeeping
  VERSION = 3
end
